### PR TITLE
LUCENE-10620: Pass the Weight to Collectors.

### PR DESCRIPTION
**Pull request by jpountz**
This allows `Collector`s to use `Weight#count` when appropriate.

See [LUCENE-10620](https://issues.apache.org/jira/browse/LUCENE-10620).

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/search/TotalHitCountCollector.java [hunk 1]
```diff
@@ -16,13 +16,17 @@
  */
 package org.apache.lucene.search;
 
+import java.io.IOException;
+import org.apache.lucene.index.LeafReaderContext;
+
 /**
  * Just counts the total number of hits. For cases when this is the only collector used, {@link
  * IndexSearcher#count(Query)} should be called instead of {@link IndexSearcher#search(Query,
```
Comment (romseygeek) [inline lucene/core/src/java/org/apache/lucene/search/TotalHitCountCollector.java:24]: I don't think this javadoc comment is accurate anymore with these changes?

File: lucene/test-framework/src/java/org/apache/lucene/tests/search/AssertingCollector.java [hunk 2]
```diff
@@ -65,4 +68,11 @@ public void collect(int doc) throws IOException {
       }
     };
   }
+
+  @Override
+  public void setWeight(Weight weight) {
+    weightSet = true;
```
Comment (romseygeek) [inline lucene/test-framework/src/java/org/apache/lucene/tests/search/AssertingCollector.java:74]: Should we assert that the Weight is only set once as well?

**Review by romseygeek — APPROVED** [review]
LGTM

**Comment by jpountz** [comment]
Thanks for looking @romseygeek. To make sure this new API would effectively have more than one use-case, I migrated `TopScoreDocCollector` and `TopFieldCollector` to it too. The immediate benefit is that collectors that pass a `totalHitsThreshold` of `Integer.MAX_VALUE` will still be able to skip non-competitive hits if the weight supports counting hits. In addition to that, I fixed some tests that were assuming that `TotalHitCountCollector` would naively iterate over matches by using a new `DummyTotalHitCountCollector` instead.

I verified that there is no performance impact on luceneutil using `wikimedium10m`:

```
                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
                        HighTerm     2374.78      (5.1%)     2297.55      (5.2%)   -3.3% ( -12% -    7%) 0.047
                         MedTerm     2795.30      (5.4%)     2704.66      (5.6%)   -3.2% ( -13% -    8%) 0.063
                    OrNotHighMed     1448.25      (3.9%)     1427.48      (4.5%)   -1.4% (  -9% -    7%) 0.286
                   OrNotHighHigh      996.35      (3.1%)      982.37      (4.6%)   -1.4% (  -8% -    6%) 0.255
                    OrHighNotMed     1898.69      (3.8%)     1876.02      (4.7%)   -1.2% (  -9% -    7%) 0.375
                      AndHighLow     1049.40      (3.3%)     1042.92      (3.8%)   -0.6% (  -7% -    6%) 0.583
                HighSloppyPhrase       21.77      (4.0%)       21.66      (4.8%)   -0.5% (  -8% -    8%) 0.716
                         LowTerm     2640.20      (6.3%)     2629.11      (4.2%)   -0.4% ( -10% -   10%) 0.803
                    OrHighNotLow     1667.62      (4.2%)     1660.75      (5.6%)   -0.4% (  -9% -    9%) 0.794
                    OrNotHighLow     1663.32      (3.0%)     1658.41      (4.2%)   -0.3% (  -7% -    7%) 0.801
                 LowSloppyPhrase       54.27      (3.1%)       54.15      (3.6%)   -0.2% (  -6% -    6%) 0.834
                   OrHighNotHigh     1259.39      (3.7%)     1257.03      (4.7%)   -0.2% (  -8% -    8%) 0.889
                 MedSloppyPhrase      115.91      (4.3%)      115.79      (6.1%)   -0.1% ( -10% -   10%) 0.952
                        PKLookup      249.41      (1.2%)      249.32      (1.5%)   -0.0% (  -2% -    2%) 0.934
                          Fuzzy2      118.47      (1.1%)      118.75      (1.2%)    0.2% (  -2% -    2%) 0.538
                         Respell       74.59      (1.1%)       74.90      (1.5%)    0.4% (  -2% -    3%) 0.323
                          IntNRQ      682.36      (2.8%)      685.81      (3.7%)    0.5% (  -5% -    7%) 0.628
                          Fuzzy1      124.32      (1.1%)      125.09      (1.1%)    0.6% (  -1% -    2%) 0.079
                       MedPhrase      623.13      (3.3%)      627.26      (3.0%)    0.7% (  -5% -    7%) 0.502
                       OrHighMed      130.02      (3.7%)      130.94      (4.2%)    0.7% (  -6% -    8%) 0.571
                       LowPhrase      110.49      (3.6%)      111.30      (2.5%)    0.7% (  -5% -    7%) 0.459
                        Wildcard       40.65      (1.6%)       40.95      (1.8%)    0.7% (  -2% -    4%) 0.167
                       OrHighLow     1092.12      (3.0%)     1101.15      (2.7%)    0.8% (  -4% -    6%) 0.360
                      AndHighMed      234.73      (4.5%)      236.77      (5.3%)    0.9% (  -8% -   11%) 0.575
                     MedSpanNear       28.83      (4.1%)       29.14      (3.3%)    1.1% (  -6% -    8%) 0.369
                     LowSpanNear       16.20      (4.2%)       16.38      (3.4%)    1.1% (  -6% -    9%) 0.363
                    HighSpanNear        7.51      (4.7%)        7.59      (3.5%)    1.1% (  -6% -    9%) 0.405
                     AndHighHigh       70.69      (5.3%)       71.60      (6.4%)    1.3% (  -9% -   13%) 0.486
                      OrHighHigh       30.64      (3.2%)       31.07      (4.3%)    1.4% (  -5% -    9%) 0.244
                      HighPhrase       22.89      (3.8%)       23.25      (3.6%)    1.6% (  -5% -    9%) 0.178
                         Prefix3      421.34      (3.5%)      430.69      (4.4%)    2.2% (  -5% -   10%) 0.078
             LowIntervalsOrdered       67.14      (4.8%)       69.35      (5.5%)    3.3% (  -6% -   14%) 0.043
            HighIntervalsOrdered        6.49      (7.8%)        6.73      (7.1%)    3.7% ( -10% -   20%) 0.112
             MedIntervalsOrdered       37.02      (7.8%)       38.45      (7.3%)    3.9% ( -10% -   20%) 0.108
           HighTermDayOfYearSort      144.92      (3.7%)      150.78      (4.6%)    4.0% (  -4% -   12%) 0.002
                      TermDTSort      204.11      (7.0%)      213.24      (7.7%)    4.5% (  -9% -   20%) 0.055
               HighTermMonthSort      154.26      (4.0%)      161.70      (4.9%)    4.8% (  -3% -   14%) 0.001
            HighTermTitleBDVSort      248.08      (3.7%)      262.32      (8.8%)    5.7% (  -6% -   18%) 0.007
```

**Comment by jpountz** [comment]
Now when collectors need to count hits too (I changed IndexSearcher's `TOTAL_HITS_THRESHOLD` to `Integer.MAX_VALUE`):

```
                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
                       OrHighLow       90.31      (6.7%)       35.49      (1.8%)  -60.7% ( -64% -  -55%) 0.000
                      OrHighHigh       40.98      (5.6%)       21.17      (2.4%)  -48.3% ( -53% -  -42%) 0.000
                    OrHighNotLow      143.61      (8.2%)       76.04      (5.4%)  -47.1% ( -56% -  -36%) 0.000
                    OrHighNotMed       88.77      (7.7%)       49.41      (5.6%)  -44.3% ( -53% -  -33%) 0.000
                   OrHighNotHigh       18.24      (7.4%)       10.59      (5.9%)  -41.9% ( -51% -  -30%) 0.000
                       OrHighMed       80.82      (5.0%)       48.18      (2.8%)  -40.4% ( -45% -  -34%) 0.000
                   OrNotHighHigh       51.35      (5.7%)       39.11      (5.6%)  -23.8% ( -33% -  -13%) 0.000
                     AndHighHigh       53.49      (1.9%)       41.97      (4.1%)  -21.5% ( -27% -  -15%) 0.000
                      AndHighMed      321.43      (2.4%)      258.39      (4.5%)  -19.6% ( -25% -  -13%) 0.000
                      AndHighLow     1777.06      (2.7%)     1474.52      (3.1%)  -17.0% ( -22% -  -11%) 0.000
                       MedPhrase      391.41      (5.9%)      332.93      (5.1%)  -14.9% ( -24% -   -4%) 0.000
                    OrNotHighMed      313.44      (6.7%)      269.25      (5.3%)  -14.1% ( -24% -   -2%) 0.000
                    OrNotHighLow     1977.65      (4.2%)     1803.88      (4.7%)   -8.8% ( -16% -    0%) 0.000
        AndHighHighDayTaxoFacets       25.28      (1.7%)       23.30      (1.9%)   -7.8% ( -11% -   -4%) 0.000
            MedTermDayTaxoFacets       79.97      (2.6%)       74.42      (3.6%)   -6.9% ( -12% -    0%) 0.000
                         Prefix3       27.72      (6.2%)       25.83      (5.2%)   -6.8% ( -17% -    4%) 0.000
                       LowPhrase      159.63      (5.0%)      148.90      (3.3%)   -6.7% ( -14% -    1%) 0.000
          OrHighMedDayTaxoFacets       19.30      (5.7%)       18.11      (4.2%)   -6.2% ( -15% -    3%) 0.000
                      HighPhrase       16.15      (5.7%)       15.30      (4.6%)   -5.2% ( -14% -    5%) 0.001
                        Wildcard       79.98      (2.3%)       76.50      (3.0%)   -4.4% (  -9% -    1%) 0.000
         AndHighMedDayTaxoFacets       72.60      (2.1%)       69.79      (1.9%)   -3.9% (  -7% -    0%) 0.000
                    HighSpanNear       44.71      (4.9%)       43.00      (4.7%)   -3.8% ( -12% -    6%) 0.012
       BrowseDayOfYearTaxoFacets       47.80      (2.0%)       46.05     (12.3%)   -3.7% ( -17% -   10%) 0.189
                          Fuzzy2      103.64      (2.1%)      100.39      (1.7%)   -3.1% (  -6% -    0%) 0.000
            BrowseDateTaxoFacets       46.13      (1.9%)       44.69     (12.0%)   -3.1% ( -16% -   10%) 0.249
     BrowseRandomLabelTaxoFacets       37.71      (2.1%)       36.53     (10.5%)   -3.1% ( -15% -    9%) 0.195
                     MedSpanNear       68.62      (3.0%)       66.69      (3.0%)   -2.8% (  -8% -    3%) 0.003
                     LowSpanNear       57.05      (3.0%)       55.49      (2.8%)   -2.7% (  -8% -    3%) 0.003
           BrowseMonthTaxoFacets       29.68      (7.3%)       28.87     (12.8%)   -2.7% ( -21% -   18%) 0.410
                          Fuzzy1      128.59      (2.2%)      125.27      (1.8%)   -2.6% (  -6% -    1%) 0.000
             LowIntervalsOrdered      219.66      (4.2%)      216.18      (3.4%)   -1.6% (  -8% -    6%) 0.184
            HighIntervalsOrdered       35.55      (5.7%)       35.03      (4.3%)   -1.5% ( -10% -    9%) 0.361
                HighSloppyPhrase        8.33     (15.0%)        8.22     (13.6%)   -1.3% ( -25% -   32%) 0.775
       BrowseDayOfYearSSDVFacets       21.93      (9.9%)       21.80      (9.3%)   -0.6% ( -17% -   20%) 0.841
           BrowseMonthSSDVFacets       23.61      (8.5%)       23.53      (8.2%)   -0.3% ( -15% -   17%) 0.904
                         Respell       77.59      (2.0%)       77.42      (2.3%)   -0.2% (  -4% -    4%) 0.740
     BrowseRandomLabelSSDVFacets       15.20      (5.5%)       15.19      (5.7%)   -0.1% ( -10% -   11%) 0.971
             MedIntervalsOrdered       43.08      (5.0%)       43.14      (4.6%)    0.1% (  -9% -   10%) 0.934
                 LowSloppyPhrase       54.27     (10.7%)       54.76     (10.1%)    0.9% ( -17% -   24%) 0.782
            BrowseDateSSDVFacets        4.21     (12.4%)        4.26     (11.8%)    1.0% ( -20% -   28%) 0.784
                        PKLookup      240.85      (2.3%)      244.63      (1.9%)    1.6% (  -2% -    5%) 0.018
                 MedSloppyPhrase       15.08      (9.3%)       15.47      (9.8%)    2.6% ( -15% -   24%) 0.384
                      TermDTSort      104.10      (2.0%)      108.13      (3.5%)    3.9% (  -1% -    9%) 0.000
           HighTermDayOfYearSort      103.69      (2.0%)      107.85      (3.5%)    4.0% (  -1% -    9%) 0.000
            HighTermTitleBDVSort      106.26      (2.3%)      112.05      (4.1%)    5.5% (   0% -   12%) 0.000
               HighTermMonthSort      210.13      (2.5%)      224.43     (12.5%)    6.8% (  -7% -   22%) 0.017
                         LowTerm      754.49     (16.4%)     2902.39     (24.5%)  284.7% ( 209% -  389%) 0.000
                         MedTerm      251.51      (3.4%)     2585.44     (48.9%)  928.0% ( 846% - 1015%) 0.000
                        HighTerm      124.40      (3.1%)     1782.93     (74.1%) 1333.3% (1217% - 1456%) 0.000
                          IntNRQ       12.42     (10.8%)      308.10    (193.8%) 2380.8% (1963% - 2899%) 0.000
```

 - IntNRQ and term queries benefit from this change the most because `Weight#count` gives the hit count up-front which then enables skipping non-competitive hits.
 - Pure disjunctions suffer the most because BS1 is no longer used since the weight doesn't know if hits will be skipped based on scores or not.
 - Other scoring queries are impacted because they need to read impacts in case the collector would like to skip based on scores.

**Comment by jpountz** [comment]
Unfortunately this is challenging to do right at the moment since the API requires the collector to tell the `ScoreMode` it needs to be able to create the `Weight`. So either the collector says it needs to evaluate all hits (`ScoreMode.COMPLETE`) and then we cannot skip hits in the case when the weight can count its hits efficiently. Or it says it doesn't (`ScoreMode.TOP_SCORES`), like the PR does at the moment and then queries get slower when the weight cannot count hits. We could fix this by moving the score mode to `LeafCollector` instead of `Collector` but this would be a big change...

**Comment by jpountz** [comment]
I reverted changes to top-docs collectors. This means this new `Collector#setWeight` API is only useful to `TotalHitCountCollector`. I've been wondering if it was worth adding a new API only for `TotalHitCountCollector` but looking at how facets use this collector, I suspect that many users set up their collectors manually instead of using `IndexSearcher#count` and do not benefit from this optimization, so maybe it's worth the increased API surface.

I was chatting about the API with @romseygeek, and we wondered if having the counting logic on `Scorable` would help. I looked into it, and it's not super practical due to the fact that `LeafCollector#setScorer` is not a place where throwing a `CollectorTerminatedException` is supported now (though this could be addressed) and this method can be called multiple times per segment, so we would need to introduce tracking to make sure that we only increment the count on the first time that `setScorer` is called on a segment. For these reasons, I would prefer moving forward with the current API on `Collector`.

**Comment by gsmiller** [comment]
> I've been wondering if it was worth adding a new API only for TotalHitCountCollector but looking at how facets use this collector, I suspect that many users set up their collectors manually instead of using IndexSearcher#count and do not benefit from this optimization, so maybe it's worth the increased API surface.

I'm not sure I'm completely following this bit. You're not suggesting that the static `FacetsCollector#search` methods could benefit from the counting optimizations right? Since the query needs to be evaluated to populate the `FacetsCollector` anyway? Is your point more that we may have users creating their own `TotalHitCountCollector` instances and passing them to `IndexSearcher#search` instead of relying on `IndexSearcher#count` when all they want is the count, following the pattern in `FacetsCollector#doSearch`? Sorry if I'm misunderstanding your point, but I got curious when I saw reference to "faceting" :).

**Comment by jpountz** [comment]
I guess I was implicitly raising the question of whether it's worth optimizing `TotalHitCountCollector` to leverage `Weight#count` given that `IndexSearcher#count` is already optimized to leverage `Weight#count`. One argument against this is that users sometimes create their collectors manually, and actually they have to if they want to be able to use `MultiCollector`. With this change, the query does indeed still have to visit every hit, but the query directly collects into the facets collector instead of going through a MultiLeafCollector that itself collects into both the `TotalHitCountCollector` and the `FacetsCollector`.

**Comment by gsmiller** [comment]
Got it, thanks @jpountz. I haven't looked at this proposed PR in much detail, so apologies in advance if this suggestion is nonsense, but if we want to optimize hit counting for users that aren't using `IndexSearcher#count`, but instead chaining collectors manually, using `TotalHitCountCollector` directly, etc., what if we did something like this?

https://github.com/gsmiller/lucene/commit/94aeb32b9e4c38075b3a6be2d2bb15b060dc0f88

**Comment by gsmiller** [comment]
@jpountz ah right. No, I don’t think it makes sense for users to have to deal with creating weights on their own (and having to consider query rewriting as well before doing so). Your approach makes total sense to me in light of this. +1 to making this change. Thanks!

**Comment by jpountz** [comment]
Thanks for taking the time to think about it @gsmiller, appreciated!

**Pull request merged**